src/elliptic.jl

```diff
@@ -0,0 +1,245 @@
+    end
+end
+
+function ellipj_smallm(u,m)
+    # [1, Sec 16.13] / https://dlmf.nist.gov/22.10.ii
+    if VERSION < v"0.7-"
+        sinu = sin(u)
+        cosu = cos(u)
+    else
+        sinu, cosu = sincos(u)
+    end
+    sn = sinu - m*(u-sinu*cosu)*cosu/4
+    cn = cosu + m*(u-sinu*cosu)*sinu/4
+    dn = 1 - m*sinu^2/2;
+    return sn,cn,dn
+end
+function ellipj_largem(u,m1)
+    # [1, Sec 16.15] / https://dlmf.nist.gov/22.10.ii
+    sinhu = sinh(u)
+    coshu = cosh(u)
+    tanhu = sinhu/coshu
+    sechu = 1/coshu
+    sn = tanhu + m1*(sinhu*coshu-u)*sechu^2/4
+    cn = sechu - m1*(sinhu*coshu-u)*tanhu*sechu/4
+    dn = sechu + m1*(sinhu*coshu+u)*tanhu*sechu/4
+    return sn,cn,dn
+end
+
+function ellipj_growm(sn,cn,dn, k)
+    # [1, Sec 16.12] / https://dlmf.nist.gov/22.7.i
+    for kk in reverse(k)
+        sn,cn,dn = (1+kk)*sn/(1+kk*sn^2),
+                       cn*dn/(1+kk*sn^2),
+                 (1-kk*sn^2)/(1+kk*sn^2)
+                 # ^ Use [1, 16.9.1]. Idea taken from [2]
+    end
+    return sn,cn,dn
+end
+function ellipj_shrinkm(sn,cn,dn, k::NTuple{N,<:Any}) where {N}
+    # [1, Sec 16.14] / https://dlmf.nist.gov/22.7.ii
+    for kk in reverse(k)
+        sn,cn,dn = (1+kk)*sn*cn/dn,
+                 (cn^2-kk*sn^2)/dn, # Use [1, 16.9.1]
+                 (cn^2+kk*sn^2)/dn  # Use [1, 16.9.1]
+    end
+    return sn,cn,dn
+end
+
+function ellipj_viasmallm(u,m,::Val{N}) where {N}
+    k,f,m = shrinkm(m,Val{N}())
+    sn,cn,dn = ellipj_smallm(u/f,m)
+    return ellipj_growm(sn,cn,dn,k)
+end
+function ellipj_vialargem(u,m,::Val{N}) where {N}
+    k,f,m1 = shrinkm(1-m,Val{N}())
+    sn,cn,dn = ellipj_largem(u/f,m1)
+    return ellipj_shrinkm(sn,cn,dn,k)
+end
+
+
+#----------------
+# Pick algorithm
+
+function ellipj_dispatch(u,m, ::Val{N}) where {N}
+    if abs(m) <= 1 && real(m) <= 0.5
+        return ellipj_viasmallm(u,m, Val{N}())
+    elseif abs(1-m) <= 1
+        return ellipj_vialargem(u,m, Val{N}())
+    elseif imag(m) == 0 && real(m) < 0
+        # [1, Sec 16.10]
+        sn,cn,dn = ellipj_dispatch(u*sqrt(1-m),-m/(1-m), Val{N}())
+        return sn/(dn*sqrt(1-m)), cn/dn, 1/dn
+    else
+        # [1, Sec 16.11]
+        sn,cn,dn = ellipj_dispatch(u*sqrt(m),1/m, Val{N}())
+        return sn/sqrt(m), dn, cn
+    end
+end
+
+Base.@pure puresqrt(x::Float64) = sqrt(x)
+Base.@pure function nsteps(m,ε)
+    i = 0
+    while abs(m) > ε
+        m = descstep(m)^2
+        i += 1
+    end
+    return i
+end
+Base.@pure nsteps(ε,::Type{<:Real}) = nsteps(0.5,ε) # Guarantees convergence in [-1,0.5]
+Base.@pure nsteps(ε,::Type{<:Complex}) = nsteps(0.5+sqrt(3)/2im,ε) # This is heuristic.
+function ellipj_nsteps(u,m)
+    # Compute the number of Landen steps required to reach machine precision.
+    # For all FloatXX types, this can be done at compile time, while for
+    # BigFloat this has to be done at runtime.
+    T = promote_type(typeof(u),typeof(m))
+    ε = puresqrt(Float64(eps(real(typeof(m)))))
```

Comment:
```suggestion
    ε = sqrt(Float64(eps(real(typeof(m)))))
```
should be sufficient in most cases.